### PR TITLE
Automatically reset window position when disconnecting secondary screen

### PR DESCRIPTION
**Pull request by tomioe**
Added a few features that I find immensely useful in my everyday use of Caprine (on Windows 10).

1. When going from multi-screen to single-screen (i.e. when undocking laptop) the main window of Caprine will sometimes be stuck off-screen (cannot even using Win+Arrow Keys to move it). Having a 'Reset Position' in the tray menu helps this problem.

2. I try to minimize my mouse movement as much as possible, but using Alt to toggle the menu and going into Exit simply isn't fast enough. I believe Command+Q is automatically bound in macOS to make an application quit, but not so on Windows, so I have added the Ctrl+Q keybinding to quit Caprine.

### CODE REVIEW

**Comment by tomioe** [comment]
For some reason I am getting an `require-atomic-updates rule is disabled but never reported` when building on `source\browser.ts:581:31` and `source\browser\conversation-list.ts:90:32`.

**Comment by sindresorhus** [comment]
> When going from multi-screen to single-screen (i.e. when undocking laptop) the main window of Caprine will sometimes be stuck off-screen (cannot even using Win+Arrow Keys to move it). Having a 'Reset Position' in the tray menu helps this problem.

I guess this is the cause: https://github.com/electron/electron/issues/10862

**Comment by sindresorhus** [comment]
> For some reason I am getting an require-atomic-updates rule is disabled but never reported when building on source\browser.ts:581:31 and source\browser\conversation-list.ts:90:32

That means you can just remove those disable comments.

**Comment by sindresorhus** [comment]
Can you try to do this automatically? You could listen to this event https://www.electronjs.org/docs/api/screen#event-display-removed and then reset the position. I wonder if after resetting the position, we could set the saved position again and it would be correct. Worth a shot.

**Comment by tomioe** [comment]
> 
> 
> > For some reason I am getting an require-atomic-updates rule is disabled but never reported when building on source\browser.ts:581:31 and source\browser\conversation-list.ts:90:32
> 
> That means you can just remove those disable comments.

[Removing those disable comments](https://github.com/sindresorhus/caprine/pull/1280/commits/fbdadf3aed5c24f83ae81cffcd7b4925bd572bd9) makes my build pass locally, but fails when [building in Travis](https://travis-ci.org/sindresorhus/caprine/builds/652109923). For some reason, I don't get the `Possible race condition...` error when building locally.

**Comment by tomioe** [comment]
> 
> 
> Can you try to do this automatically? You could listen to this event https://www.electronjs.org/docs/api/screen#event-display-removed and then reset the position. I wonder if after resetting the position, we could set the saved position again and it would be correct. Worth a shot.

Will give this a shot, would definitely be smoother to have it happen automatically.

**Comment by tomioe** [comment]
> 
> 
> Can you try to do this automatically? You could listen to this event https://www.electronjs.org/docs/api/screen#event-display-removed and then reset the position. I wonder if after resetting the position, we could set the saved position again and it would be correct. Worth a shot.

I've tried an implementing this in [9f6e1ad](https://github.com/sindresorhus/caprine/pull/1280/commits/9f6e1adb3b92e705f7881bd325ee076edb3d9b05). It can be tested in Windows (if using two screens), by moving the Caprine main window to the secondary screen and setting the "Project" (Win+P) setting to "PC Screen Only". The Caprine window position should move to the same position on the primary screen as it had on the secondary screen.

File: source/index.ts [hunk 1]
```diff
@@ -92,6 +92,14 @@ app.on('second-instance', () => {
 	}
 });
 
+// Main window should move to the same position as when it was another screen
+app.on('ready', () => {
+	electronScreen.on('display-removed', () => {
+		const oldPosition = mainWindow.getPosition();
```
Comment (sindresorhus) [inline source/index.ts]:
```suggestion
		const [x, y] = mainWindow.getPosition();
```

File: source/index.ts [hunk 2]
```diff
@@ -92,6 +92,14 @@ app.on('second-instance', () => {
 	}
 });
 
+// Main window should move to the same position as when it was another screen
```
Comment (sindresorhus) [inline source/index.ts]: This sentence is not grammatically correct.

File: source/tray.ts [hunk 3]
```diff
@@ -73,6 +73,14 @@ export default {
...
 				}
 			},
+			{
+				label: 'Reset Position',
+				visible: !is.macos,
+				click() {
+					win.setPosition(0, 0);
+				}
+
+			},
```
Comment (sindresorhus) [inline source/tray.ts]: This is no longer needed.

File: source/browser.ts [hunk 4]
```diff
@@ -573,7 +573,7 @@ document.addEventListener('DOMContentLoaded', async () => {
 	// Prevent flash of white on startup when in dark mode
 	// TODO: find a CSS-only solution
 	if (!is.macos && config.get('darkMode')) {
-		// eslint-disable-next-line require-atomic-updates
+		/* eslint-disable-next-line require-atomic-updates */
```
Comment (sindresorhus) [inline source/browser.ts]: Don't do unrelated changes.

**Comment by sindresorhus** [comment]
> I believe Command+Q is automatically bound in macOS to make an application quit, but not so on Windows, so I have added the Ctrl+Q keybinding to quit Caprine.

Is that really a Windows convention? Do other Windows app support that. Because the menu item is named "Exit" on windows, so "Q" doesn't really make sense.

**Comment by tomioe** [comment]
> 
> 
> > I believe Command+Q is automatically bound in macOS to make an application quit, but not so on Windows, so I have added the Ctrl+Q keybinding to quit Caprine.
> 
> Is that really a Windows convention? Do other Windows app support that. Because the menu item is named "Exit" on windows, so "Q" doesn't really make sense.

Huh, I must've picked up that bad habit over the years... I just assumed it was a conventional shortcut, due to how closely it is to Cmd+Q, but I was wrong: https://support.microsoft.com/en-us/help/12445/windows-keyboard-shortcuts

It could be argued, that although uncommon, it is still implemented in some applications (i.e. Firefox uses `Ctrl+Shift+Q`), while others seem to favor the more idiomatic `Alt+F, X`:https://www.ghacks.net/2018/10/22/google-retires-ctrl-shift-q-in-chrome-to-exit-web-browser/

Currently `Alt+F, X` does not work in Carpine, since underlined letters for keyboard navigation have not been implemented. Perhaps this is a better suited goal of this PR?

**Comment by sindresorhus** [comment]
> Currently Alt+F, X does not work in Carpine, since underlined letters for keyboard navigation have not been implemented. Perhaps this is a better suited goal of this PR?

Yes, but that's really out of the scope of this PR and should be a separate PR.

File: source/menu.ts [hunk 5]
```diff
@@ -744,23 +745,27 @@ ${debugInfo()}`;
...
 				{
+					label: 'E&xit',
 					role: 'quit'
 				}
 			]
 		},
 		{
-			role: 'editMenu'
+			role: 'editMenu',
+			label: '&Edit'
```
Comment (sindresorhus) [inline source/menu.ts]: I think you should rather open an issue on Electron about this. I'm already using the `role`, which should give the correct automatic behavior. If `&Edit` (and others) is indeed correct on Windows, it should be handled by Electron.

File: source/index.ts [hunk 6]
```diff
@@ -92,6 +92,14 @@ app.on('second-instance', () => {
 	}
 });
 
+// Main window should move to the same position as it had on another screen
```
Comment (sindresorhus) [inline source/index.ts]:
```suggestion
// Preserves the window position when a display is removed and Caprine is moved to a different screen.
```

**Comment by sindresorhus** [comment]
The PR title needs to be updated.

File: source/menu.ts (hunk 5 above)
Comment (tomioe) [reply source/menu.ts]: It seems like the default roles do not generate these accelerators: https://github.com/electron/electron/blob/a5f87cee6b79e2ed38fbfe7b8af26a9de0649658/lib/browser/api/menu-item-roles.js#L187-L193

The documentation also calls for using '&', which I believe is because it does not want to be too presumptuous about application keyboard shortcuts: https://www.electronjs.org/docs/api/menu#menusetapplicationmenumenu

But as you said, this is out of scope for this PR, I will create a separate PR for this.

File: source/index.ts (hunk 6 above)
Comment (tomioe) [reply source/index.ts]: Appreciated, I was struggling with wording on this one ...

File: source/menu.ts (hunk 5 above)
Comment (sindresorhus) [reply source/menu.ts]: I still think it's worth opening an Electron issue. If it doesn't make sense to add default keyboard shortcuts for the defined role menus, I'm not sure why apps should do it at all.

File: source/menu.ts [hunk 7]
```diff
@@ -756,7 +756,6 @@ ${debugInfo()}`;
 			submenu: viewSubmenu
 		},
 		{
-			label: 'Conversation',
```
Comment (sindresorhus) [inline source/menu.ts]: ?

Comment (tomioe) [reply source/menu.ts]: Absolute blunder on my part, apologies...
